README.md
Plugin: check_neon
==================

This is a ICINGA/NAGIOS plugin for the Neon 100 device which is manufactured by Sensormetrix.

Behind the scenes
-----------------

The plugin just fech the xml data from the device and probe for thresholds.

Installation
------------

Installation ist just simple. You need Python >= 2.6 to run this

### Python Setuptools

To install nagiosplugins python package you need Python Setuptools. Just make sure to have that installed. For RHEL or
CentOS you can simply run

    # yum install python-setuptools.noarch

### nagiosplugin Module

You need the Python nagiosplugin package. More information can be found on

* [Package website](http://pythonhosted.org/nagiosplugin/index.html)
* [PyPi repository](https://pypi.python.org/pypi/nagiosplugin/)

Extract the package, go into the directory and type

    # wget https://pypi.python.org/packages/source/n/nagiosplugin/nagiosplugin-1.2.1.tar.gz#md5=d81c724525e8e8b290d17046109e71d2
    # cd nagiosplugin-1.2.1
    # python setup.py install

### check_neon

Move the check_neon.py script to your favorite ICINGA resource directory, e.g.

    # cp check_neon.py /usr/local/icinga/libexec

Usage
-----

First, you should try that everything is working:

    # ./check_neon.py --host=X.X.X.X  --info
    Device information:
    ---------------------
    dhcp                : 1
    dst                 : 0
    gmtoffset           : 0
    ipaddress           : X.X.X.X
    mac                 : FF:FF:FF:FF:FF:FF
    netmask             : 255.255.0.0
    powerup             : 21 days, 3 hrs, 49 min, 51 sec
    productname         : Neon110
    serialnumber        : 036030
    time                : 03:30:33 PM
    unitname            : Neon110
    version             : V1.4

    Sensor humidity
    ---------------------
    alarm               : 0
    highalarm           : 80
    lowalarm            : 20
    unit                : %RH
    value               : 30.67

    Sensor temperature
    ---------------------
    alarm               : 0
    highalarm           : 25
    lowalarm            : 0
    unit                : C
    value               : 21.14

Probe device
------------

Just type this:

    # ./check_neon.py --host=10.0.10.81
    NEON OK - temperature is 21.14 C, humidity is 30.52 %RH
    | humidity=30.52;20:80;20:80 temperature=21.14;25;25

We check temperature and humidity always together. If you to not set critical or warning thresholds we'll use the
Hi- and LoAlarm from webinterface

Probe with own thresholds:

    # ./check_neon.py --host=10.0.10.81 -c 20
    NEON CRITICAL - temperature is 21.15 (outside range 0:20)
    critical: temperature is 21.15 (outside range 0:20)
    critical: humidity is 30.33 (outside range 0:20)
    | humidity=30.33;20:80;20 temperature=21.15;25;20

### Program switches:

Simple usage: **./check_neon.py --host=<hostname | ip address>**

**-h | --help**

Displays a small help screen and exit

**-H | --host=<STRING>**

Device target. This could be an ip adress of a dns hostname

**-w | --warning=<RANGE>**

Threshold for warning state. Have a look on the nagios plugin development guidelines
[here](https://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT)

**-c | --critical=<RANGE>**

Threshold for critical state. Have a look on the nagios plugin development guidelines
[here](https://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT)

**-T | --timeout=<int>**

Timeout in seconds for http requests to get the data from device

Copyright
---------

Copyright (c) 2014 NETWAYS GmbH

[www.netways.de](http://www.netways.de)

info@netways.de

Where to buy
------------

Product can be found here:
[NETWAYS Shop](http://shop.netways.de/ueberwachung/sequoia/messgerate/neon-110-netzwerksensor-fur-temperatur-und-luftfeuchtigkeit.html)

Bugs and repository
-------------------

For chrous of praise or complaints you can go here:

* [https://git.netways.org/plugins/check_neon](https://git.netways.org/plugins/check_neon)
* [https://www.netways.org/projects/plugins](https://www.netways.org/projects/plugins)

Native git access: git://git.netways.org/plugins/check_neon.git

NEWS.md
News
====

Product can be found here: [NETWAYS Shop](http://shop.netways.de/ueberwachung/sequoia/messgerate/neon-110-netzwerksensor-fur-temperatur-und-luftfeuchtigkeit.html)

check_neon.py
#!/usr/bin/env python

# COPYRIGHT
#
# This software is Copyright (c)  2011 NETWAYS GmbH, Gunnar Beutner
#                                 <support@netways.de>
#
# (Except where explicitly superseded by other copyright notices)
#
# LICENSE
#
# This work is made available to you under the terms of Version 2 of
# the GNU General Public License. A copy of that license should have
# been provided with this software, but in any event can be snarfed
# from http://www.fsf.org.
#
# This work is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301 or visit their web page on the internet at
# http://www.fsf.org.
#
#
# CONTRIBUTION SUBMISSION POLICY:
#
# (The following paragraph is not intended to limit the rights granted
# to you to modify and distribute this software under the terms of
# the GNU General Public License and is only of importance to you if
# you choose to contribute your changes and enhancements to the
# community by submitting them to NETWAYS GmbH.)
#
# By intentionally submitting any modifications, corrections or
# derivatives to this work, or any other work intended for use with
# this Software, to NETWAYS GmbH, you confirm that
# you are the copyright holder for those contributions and you grant
# NETWAYS GmbH a nonexclusive, worldwide, irrevocable,
# royalty-free, perpetual, license to use, copy, create derivative
# works based on those contributions, and sublicense and distribute
# those contributions and any derivatives thereof.
#
# Nagios and the Nagios logo are registered trademarks of Ethan Galstad.

import sys
import nagiosplugin
import urllib2
from optparse import OptionParser
from xml.dom import minidom

__version__ = '0.1'

class DataContainer:
    """ Container class to fetch data from device
    """
    def __init__(self, host, timeout=4):
        """ Create a new container
            Args:
                host (str): IP or hostname
                timeout (int): Timeout in seconds
        """
        self.url = 'http://' + host + '/values.xml'
        self.timeout = timeout

    def getText(self, nodelist):
        """ Convert XML text nodes into strings
            Args:
                nodelist (DOM nodelist): A node with sub nodes
        """
        rc = []
        for node in nodelist:
            if node.nodeType == node.TEXT_NODE:
                rc.append(node.data)
        return ((''.join(rc)).strip()).encode('ascii')

    def parse_nodes(self, target, nodes):
        """ Parse nodes into a dict
            Args:
                target (dict): Target list (for recursive parsing)
                nodes (DOM nodelist: List of elements to parse
        """
        for node in nodes:
            if node.nodeType == minidom.Node.ELEMENT_NODE:
                lowername = (node.nodeName.lower()).encode('ascii')
                if node.childNodes.length == 1:
                    try:
                        target[lowername] = self.getText(node.childNodes)
                    except AttributeError:
                        pass
                else:
                    target[lowername] = {}
                    self.parse_nodes(target[lowername], node.childNodes)

    def get_values(self):
        """ Return a dict of data from device
        """
        f = urllib2.urlopen(self.url, None, self.timeout)
        document = minidom.parse(f)
        root = document.getElementsByTagName('root')[0]
        target = {}
        self.parse_nodes(target, root.childNodes)
        return target

    def print_processor(self, values):
        """ Helper to print values
            Args:
                values (dict)
        """
        later = {}
        for key,value in sorted(values.items()):
            if type(value) is dict:
                later[key] = value
            else:
                print('%-20s: %s' % (key, value,))

        for key,value in sorted(later.items()):
            print("")
            print("Sensor " + key)
            print('-' * 21)
            self.print_processor(value)

    def print_values(self):
        """ Print device information to stdout
        """
        print('Device information:')
        print('-' * 21)
        self.print_processor(self.get_values())

class Neon(nagiosplugin.Resource):
    """ Resource for nagiosplugin
    """
    def __init__(self, temp, hum):
        """ Create a new Neon resource
            Args:
                temp (dict): Dict of temperature values
                hum (dict): Dict of humidity values
        """
        self.temp = temp
        self.hum = hum
    def probe(self):
        """ Probe for plugin values
        """
        yield nagiosplugin.Metric('temperature', float(self.temp['value']))
        yield nagiosplugin.Metric('humidity', float(self.hum['value']))

class NeonSummary(nagiosplugin.Summary):
    """ Building some text for the output
    """
    def __init__(self, temp, hum):
        """ Create a new summary object
            Args
                temp (dict): Dict of temperature values
                hum (dict): Dict of humidity values
        """
        self.temp = temp
        self.hum = hum
    def ok(self, results):
        """ Text message for okay results
            Args:
                resource (nagiosplugin ResourceContainer): Results
        """
        return '{0} {1}, {2} {3}'.format(
            str(results['temperature']),
            self.temp['unit'],
            str(results['humidity']),
            self.hum['unit'],)

@nagiosplugin.guarded
def main():
    """ Main execution function
    """
    parser = OptionParser(usage='%prog --host=<ip|hostname>')

    parser.add_option('-H', '--host', dest='hostname',
                      help='Hostname of the device')

    parser.add_option('-I', '--info', dest='info', action='store_true',
                      help='Display device information')

    parser.add_option('-w', '--warning', dest='warning', metavar='RANGE',
                      help='Warning threshold (optional)')

    parser.add_option('-c', '--critical', dest='critical', metavar='RANGE',
                      help='Critical threshold (optional)')

    parser.add_option('-T', '--timeout', dest='timeout', default=4, type='int',
                      help='Timeout for http requests')

    (options, args) = parser.parse_args()

    if options.hostname is None:
        parser.error('Option --host is mandatory')

    container = DataContainer(options.hostname, options.timeout)

    if options.info:
        container.print_values()
        return 3

    try:
        values = container.get_values()
    except urllib2.URLError as e:
        raise nagiosplugin.CheckError(e.reason)

    temperature_warn = None
    temperature_critical = None
    humidity_warn = None
    humidity_critical = None

    # Using default thresholds configured in the web interface
    if not options.warning:
        temperature_warn = values['temperature']['lowalarm'] + ':' + values['temperature']['highalarm']
        humidity_warn = values['humidity']['lowalarm'] + ':' + values['humidity']['highalarm']
    else:
        temperature_warn = options.warning
        humidity_warn = options.warning

    if not options.critical:
        temperature_critical = values['temperature']['lowalarm'] + ':' + values['temperature']['highalarm']
        humidity_critical = values['humidity']['lowalarm'] + ':' + values['humidity']['highalarm']
    else:
        temperature_critical = options.critical
        humidity_critical = options.critical

    check = nagiosplugin.Check(
        Neon(values['temperature'], values['humidity']),
        nagiosplugin.ScalarContext('temperature',
                                   temperature_warn,
                                   temperature_critical),
        nagiosplugin.ScalarContext('humidity',
                                   humidity_warn,
                                   humidity_critical),
        NeonSummary(values['temperature'], values['humidity']))

    return check.main()

if __name__ == '__main__':
    sys.exit(main())